router/batchrouter/asyncdestinationmanager/bing-ads/offline-conversions/bulk_uploader.go

```diff
@@ -0,0 +1,344 @@
+package bingads_offline_conversions
+
+import (
+	"encoding/json"
+	stdjson "encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/bing-ads-go-sdk/bingads"
+	"github.com/rudderlabs/rudder-go-kit/bytesize"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/router/batchrouter/asyncdestinationmanager/common"
+	router_utils "github.com/rudderlabs/rudder-server/router/utils"
+)
+
+func NewBingAdsBulkUploader(destName string, service bingads.BulkServiceI, client *Client) *BingAdsBulkUploader {
+	return &BingAdsBulkUploader{
+		destName:      destName,
+		service:       service,
+		logger:        logger.NewLogger().Child("batchRouter").Child("AsyncDestinationManager").Child("BingAds").Child("BingAdsBulkUploader"),
+		client:        *client,
+		fileSizeLimit: common.GetBatchRouterConfigInt64("MaxUploadLimit", destName, 100*bytesize.MB),
+		eventsLimit:   common.GetBatchRouterConfigInt64("MaxEventsLimit", destName, 1000),
+	}
+}
+
+/*
+	Microsoft Docs: https://learn.microsoft.com/en-us/advertising/bulk-service/offline-conversion
+
+returns: A string of object in form
+
+	{
+		message:{
+			Action: "insert", "update", "delete",
+			fields: {}
+		},
+		metadata:{
+			jobId: "job_id"
+		}
+	}
+*/
+func (b *BingAdsBulkUploader) Transform(job *jobsdb.JobT) (string, error) {
+	// Unmarshal the JSON raw message into the record struct
+	payload := string(job.EventPayload)
+	var event Record
+	var fields map[string]interface{}
+	err := json.Unmarshal(job.EventPayload, &event)
+	if err != nil {
+		fmt.Println("Error during unmarshalling event:", err)
+		return payload, err
+	}
+	err = json.Unmarshal(event.Fields, &fields)
+	if err != nil {
+		fmt.Println("Error during unmarshalling event.fields:", err)
+		return payload, err
+	}
+	// validate for conversion time mscklid and conversion name
+	generalRequiredFields := []string{"microsoftClickId", "conversionName", "conversionTime"}
+	for _, field := range generalRequiredFields {
+		err := validateField(fields, field)
+		if err != nil {
+			return payload, err
+		}
+	}
+	if event.Action != "insert" {
+		// validate for adjusted time
+		err := validateField(fields, "conversionAdjustedTime")
+		if err != nil {
+			return payload, err
+		}
+		if event.Action == "update" {
+			// validate for Adjustment Value
+			err := validateField(fields, "conversionValue")
+			if err != nil {
+				return payload, err
+			}
+		}
+	}
```

Comment:
```suggestion
	if event.Action != "insert" {
		// validate for adjusted time
		err := validateField(fields, "conversionAdjustedTime")
		if err != nil {
			return payload, err
		}
	}
	if event.Action == "update" {
		// validate for Adjustment Value
		err := validateField(fields, "conversionValue")
		if err != nil {
			return payload, err
		}
	}
```
[minor] reducing nesting to improve reliability

Reply:
Incorporated